From 79b68019ff4b17c4b80fd2c6c725071a050559ca Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Mon, 21 Jul 2025 18:07:21 -0400
Subject: [PATCH] Issue 6893 - Log user that is updated during password modify
 extended operation

Description:

When a user's password is updated via an extended operation (password modify
plugin) we only log the bind DN and not what user was updated. While "internal
operation" logging will display the the user it should be logged by the default
logging level.

Add access logging using "EXT_INFO" for the old logging format, and
"EXTENDED_OP_INFO" for json logging where we display the bind dn, target
dn, and message.

Relates: https://github.com/389ds/389-ds-base/issues/6893

Reviewed by: spichugi & tbordaz(Thanks!!)
---
 .../logging/access_json_logging_test.py       | 98 +++++++++++++++----
 ldap/servers/slapd/accesslog.c                | 47 +++++++++
 ldap/servers/slapd/passwd_extop.c             | 69 +++++++------
 ldap/servers/slapd/slapi-private.h            |  1 +
 4 files changed, 169 insertions(+), 46 deletions(-)

diff --git a/dirsrvtests/tests/suites/logging/access_json_logging_test.py b/dirsrvtests/tests/suites/logging/access_json_logging_test.py
index f0dc861a7..699bd8c4d 100644
--- a/dirsrvtests/tests/suites/logging/access_json_logging_test.py
+++ b/dirsrvtests/tests/suites/logging/access_json_logging_test.py
@@ -11,7 +11,7 @@ import os
 import time
 import ldap
 import pytest
-from lib389._constants import DEFAULT_SUFFIX, PASSWORD, LOG_ACCESS_LEVEL
+from lib389._constants import DEFAULT_SUFFIX, PASSWORD, LOG_ACCESS_LEVEL, DN_DM
 from lib389.properties import TASK_WAIT
 from lib389.topologies import topology_m2 as topo_m2
 from lib389.idm.group import Groups
@@ -548,22 +548,6 @@ def test_access_json_format(topo_m2, setup_test):
                              "2.16.840.1.113730.3.4.3",
                              "LDAP_CONTROL_PERSISTENTSEARCH")
 
-    #
-    # Extended op
-    #
-    log.info("Test EXTENDED_OP")
-    event = get_log_event(inst, "EXTENDED_OP", "oid",
-                          "2.16.840.1.113730.3.5.12")
-    assert event is not None
-    assert event['oid_name'] == "REPL_START_NSDS90_REPLICATION_REQUEST_OID"
-    assert event['name'] == "replication-multisupplier-extop"
-
-    event = get_log_event(inst, "EXTENDED_OP", "oid",
-                          "2.16.840.1.113730.3.5.5")
-    assert event is not None
-    assert event['oid_name'] == "REPL_END_NSDS50_REPLICATION_REQUEST_OID"
-    assert event['name'] == "replication-multisupplier-extop"
-
     #
     # TLS INFO/TLS CLIENT INFO
     #
@@ -579,7 +563,8 @@ def test_access_json_format(topo_m2, setup_test):
         'sn': RDN_TEST_USER,
         'uidNumber': '1000',
         'gidNumber': '2000',
-        'homeDirectory': f'/home/{RDN_TEST_USER}'
+        'homeDirectory': f'/home/{RDN_TEST_USER}',
+        'userpassword': 'password'
     })
 
     ssca_dir = inst.get_ssca_dir()
@@ -646,6 +631,83 @@ def test_access_json_format(topo_m2, setup_test):
     assert event['msg'] == "failed to map client certificate to LDAP DN"
     assert event['err_msg'] == "Certificate couldn't be mapped to an ldap entry"
 
+    #
+    # Extended op
+    #
+    log.info("Test EXTENDED_OP")
+    event = get_log_event(inst, "EXTENDED_OP", "oid",
+                          "2.16.840.1.113730.3.5.12")
+    assert event is not None
+    assert event['oid_name'] == "REPL_START_NSDS90_REPLICATION_REQUEST_OID"
+    assert event['name'] == "replication-multisupplier-extop"
+
+    event = get_log_event(inst, "EXTENDED_OP", "oid",
+                          "2.16.840.1.113730.3.5.5")
+    assert event is not None
+    assert event['oid_name'] == "REPL_END_NSDS50_REPLICATION_REQUEST_OID"
+    assert event['name'] == "replication-multisupplier-extop"
+
+    #
+    # Extended op info
+    #
+    log.info("Test EXTENDED_OP_INFO")
+    OLD_PASSWD = 'password'
+    NEW_PASSWD = 'newpassword'
+
+    assert inst.simple_bind_s(DN_DM, PASSWORD)
+
+    assert inst.passwd_s(user.dn, OLD_PASSWD, NEW_PASSWD)
+    event = get_log_event(inst, "EXTENDED_OP_INFO", "name",
+                          "passwd_modify_plugin")
+    assert event is not None
+    assert event['bind_dn'] == "cn=directory manager"
+    assert event['target_dn'] == user.dn.lower()
+    assert event['msg'] == "success"
+
+    # Test no such object
+    BAD_DN = user.dn + ",dc=not"
+    with pytest.raises(ldap.NO_SUCH_OBJECT):
+        inst.passwd_s(BAD_DN, OLD_PASSWD, NEW_PASSWD)
+
+    event = get_log_event(inst, "EXTENDED_OP_INFO", "target_dn", BAD_DN)
+    assert event is not None
+    assert event['bind_dn'] == "cn=directory manager"
+    assert event['target_dn'] == BAD_DN.lower()
+    assert event['msg'] == "No such entry exists."
+
+    # Test invalid old password
+    with pytest.raises(ldap.INVALID_CREDENTIALS):
+        inst.passwd_s(user.dn, "not_the_old_pw", NEW_PASSWD)
+    event = get_log_event(inst, "EXTENDED_OP_INFO", "err", 49)
+    assert event is not None
+    assert event['bind_dn'] == "cn=directory manager"
+    assert event['target_dn'] == user.dn.lower()
+    assert event['msg'] == "Invalid oldPasswd value."
+
+    # Test user without permissions
+    user2 = users.create(properties={
+        'uid': RDN_TEST_USER + "2",
+        'cn': RDN_TEST_USER + "2",
+        'sn': RDN_TEST_USER + "2",
+        'uidNumber': '1001',
+        'gidNumber': '2001',
+        'homeDirectory': f'/home/{RDN_TEST_USER + "2"}',
+        'userpassword': 'password'
+    })
+    inst.simple_bind_s(user2.dn, 'password')
+    with pytest.raises(ldap.INSUFFICIENT_ACCESS):
+        inst.passwd_s(user.dn, NEW_PASSWD, OLD_PASSWD)
+    event = get_log_event(inst, "EXTENDED_OP_INFO", "err", 50)
+    assert event is not None
+    assert event['bind_dn'] == user2.dn.lower()
+    assert event['target_dn'] == user.dn.lower()
+    assert event['msg'] == "Insufficient access rights"
+
+
+    # Reset bind
+    inst.simple_bind_s(DN_DM, PASSWORD)
+
+
 
 if __name__ == '__main__':
     # Run isolated
diff --git a/ldap/servers/slapd/accesslog.c b/ldap/servers/slapd/accesslog.c
index 072ace203..46228d4a1 100644
--- a/ldap/servers/slapd/accesslog.c
+++ b/ldap/servers/slapd/accesslog.c
@@ -1113,6 +1113,53 @@ slapd_log_access_extop(slapd_log_pblock *logpb)
     return rc;
 }
 
+/*
+ * Extended operation information
+ *
+ * int32_t log_format
+ * time_t conn_time
+ * uint64_t conn_id
+ * int32_t op_id
+ * const char *name
+ * const char *bind_dn
+ * const char *tartet_dn
+ * const char *msg
+ */
+int32_t
+slapd_log_access_extop_info(slapd_log_pblock *logpb)
+{
+    int32_t rc = 0;
+    char *msg = NULL;
+    json_object *json_obj = NULL;
+
+    if ((json_obj = build_base_obj(logpb, "EXTENDED_OP_INFO")) == NULL) {
+        return rc;
+    }
+
+    if (logpb->name) {
+        json_object_object_add(json_obj, "name", json_obj_add_str(logpb->name));
+    }
+    if (logpb->target_dn) {
+        json_object_object_add(json_obj, "target_dn", json_obj_add_str(logpb->target_dn));
+    }
+    if (logpb->bind_dn) {
+        json_object_object_add(json_obj, "bind_dn", json_obj_add_str(logpb->bind_dn));
+    }
+    if (logpb->msg) {
+        json_object_object_add(json_obj, "msg", json_obj_add_str(logpb->msg));
+    }
+    json_object_object_add(json_obj, "err", json_object_new_int(logpb->err));
+
+    /* Convert json object to string and log it */
+    msg = (char *)json_object_to_json_string_ext(json_obj, logpb->log_format);
+    rc = slapd_log_access_json(msg);
+
+    /* Done with JSON object - free it */
+    json_object_put(json_obj);
+
+    return rc;
+}
+
 /*
  * Sort
  *
diff --git a/ldap/servers/slapd/passwd_extop.c b/ldap/servers/slapd/passwd_extop.c
index 4bb60afd6..69bb3494c 100644
--- a/ldap/servers/slapd/passwd_extop.c
+++ b/ldap/servers/slapd/passwd_extop.c
@@ -465,12 +465,14 @@ passwd_modify_extop(Slapi_PBlock *pb)
     BerElement *response_ber = NULL;
     Slapi_Entry *targetEntry = NULL;
     Connection *conn = NULL;
+    Operation *pb_op = NULL;
     LDAPControl **req_controls = NULL;
     LDAPControl **resp_controls = NULL;
     passwdPolicy *pwpolicy = NULL;
     Slapi_DN *target_sdn = NULL;
     Slapi_Entry *referrals = NULL;
-    /* Slapi_DN sdn; */
+    Slapi_Backend *be = NULL;
+    int32_t log_format = config_get_accesslog_log_format();
 
     slapi_log_err(SLAPI_LOG_TRACE, "passwd_modify_extop", "=>\n");
 
@@ -647,7 +649,7 @@ parse_req_done:
     }
     dn = slapi_sdn_get_ndn(target_sdn);
     if (dn == NULL || *dn == '\0') {
-        /* Refuse the operation because they're bound anonymously */
+        /* Invalid DN - refuse the operation */
         errMesg = "Invalid dn.";
         rc = LDAP_INVALID_DN_SYNTAX;
         goto free_and_return;
@@ -724,14 +726,19 @@ parse_req_done:
         ber_free(response_ber, 1);
     }
 
-    slapi_pblock_set(pb, SLAPI_ORIGINAL_TARGET, (void *)dn);
+    slapi_pblock_get(pb, SLAPI_OPERATION, &pb_op);
+    if (pb_op == NULL) {
+        slapi_log_err(SLAPI_LOG_ERR, "passwd_modify_extop", "pb_op is NULL\n");
+        goto free_and_return;
+    }
 
+    slapi_pblock_set(pb, SLAPI_ORIGINAL_TARGET, (void *)dn);
     /* Now we have the DN, look for the entry */
     ret = passwd_modify_getEntry(dn, &targetEntry);
     /* If we can't find the entry, then that's an error */
     if (ret) {
         /* Couldn't find the entry, fail */
-        errMesg = "No such Entry exists.";
+        errMesg = "No such entry exists.";
         rc = LDAP_NO_SUCH_OBJECT;
         goto free_and_return;
     }
@@ -742,30 +749,18 @@ parse_req_done:
         leak any useful information to the client such as current password
         wrong, etc.
       */
-    Operation *pb_op = NULL;
-    slapi_pblock_get(pb, SLAPI_OPERATION, &pb_op);
-    if (pb_op == NULL) {
-        slapi_log_err(SLAPI_LOG_ERR, "passwd_modify_extop", "pb_op is NULL\n");
-        goto free_and_return;
-    }
-
     operation_set_target_spec(pb_op, slapi_entry_get_sdn(targetEntry));
     slapi_pblock_set(pb, SLAPI_REQUESTOR_ISROOT, &pb_op->o_isroot);
 
-    /* In order to perform the access control check , we need to select a backend (even though
-     * we don't actually need it otherwise).
-     */
-    {
-        Slapi_Backend *be = NULL;
-
-        be = slapi_mapping_tree_find_backend_for_sdn(slapi_entry_get_sdn(targetEntry));
-        if (NULL == be) {
-            errMesg = "Failed to find backend for target entry";
-            rc = LDAP_OPERATIONS_ERROR;
-            goto free_and_return;
-        }
-        slapi_pblock_set(pb, SLAPI_BACKEND, be);
+    /* In order to perform the access control check, we need to select a backend (even though
+     * we don't actually need it otherwise). */
+    be = slapi_mapping_tree_find_backend_for_sdn(slapi_entry_get_sdn(targetEntry));
+    if (NULL == be) {
+        errMesg = "Failed to find backend for target entry";
+        rc = LDAP_NO_SUCH_OBJECT;
+        goto free_and_return;
     }
+    slapi_pblock_set(pb, SLAPI_BACKEND, be);
 
     /* Check if the pwpolicy control is present */
     slapi_pblock_get(pb, SLAPI_PWPOLICY, &need_pwpolicy_ctrl);
@@ -797,10 +792,7 @@ parse_req_done:
     /* Check if password policy allows users to change their passwords.  We need to do
      * this here since the normal modify code doesn't perform this check for
      * internal operations. */
-
-    Connection *pb_conn;
-    slapi_pblock_get(pb, SLAPI_CONNECTION, &pb_conn);
-    if (!pb_op->o_isroot && !pb_conn->c_needpw && !pwpolicy->pw_change) {
+    if (!pb_op->o_isroot && !conn->c_needpw && !pwpolicy->pw_change) {
         if (NULL == bindSDN) {
             bindSDN = slapi_sdn_new_normdn_byref(bindDN);
         }
@@ -848,6 +840,27 @@ free_and_return:
     slapi_log_err(SLAPI_LOG_PLUGIN, "passwd_modify_extop",
                   "%s\n", errMesg ? errMesg : "success");
 
+    if (dn) {
+        /* Log the target ndn (if we have a target ndn) */
+        if (log_format != LOG_FORMAT_DEFAULT) {
+            /* JSON logging */
+            slapd_log_pblock logpb = {0};
+            slapd_log_pblock_init(&logpb, log_format, pb);
+            logpb.name = "passwd_modify_plugin";
+            logpb.target_dn = dn;
+            logpb.bind_dn = bindDN;
+            logpb.msg = errMesg ? errMesg : "success";
+            logpb.err = rc;
+            slapd_log_access_extop_info(&logpb);
+        } else {
+            slapi_log_access(LDAP_DEBUG_STATS,
+                    "conn=%" PRIu64 " op=%d EXT_INFO name=\"passwd_modify_plugin\" bind_dn=\"%s\" target_dn=\"%s\" msg=\"%s\" rc=%d\n",
+                    conn ? conn->c_connid : -1, pb_op ? pb_op->o_opid : -1,
+                    bindDN ? bindDN : "", dn,
+                    errMesg ? errMesg : "success", rc);
+        }
+    }
+
     if ((rc == LDAP_REFERRAL) && (referrals)) {
         send_referrals_from_entry(pb, referrals);
     } else {
diff --git a/ldap/servers/slapd/slapi-private.h b/ldap/servers/slapd/slapi-private.h
index da232ae2f..e9abf8b75 100644
--- a/ldap/servers/slapd/slapi-private.h
+++ b/ldap/servers/slapd/slapi-private.h
@@ -1652,6 +1652,7 @@ int32_t slapd_log_access_vlv(slapd_log_pblock *logpb);
 int32_t slapd_log_access_entry(slapd_log_pblock *logpb);
 int32_t slapd_log_access_referral(slapd_log_pblock *logpb);
 int32_t slapd_log_access_extop(slapd_log_pblock *logpb);
+int32_t slapd_log_access_extop_info(slapd_log_pblock *logpb);
 int32_t slapd_log_access_sort(slapd_log_pblock *logpb);
 int32_t slapd_log_access_tls(slapd_log_pblock *logpb);
 int32_t slapd_log_access_tls_client_auth(slapd_log_pblock *logpb);
-- 
2.49.0

